File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter12/GroupBySex/GroupBySex/StudentGroups.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using ElPasoHighSchool;

namespace GroupBySex
{
    public class StudentGroups : INotifyPropertyChanged
    {
        StudentBodyPresenter presenter;
        ObservableCollection<StudentGroup> groups = new ObservableCollection<StudentGroup>();

        public event PropertyChangedEventHandler PropertyChanged;

        public StudentBodyPresenter Source
        {
            set
            {
                if (value != null)
                {
                    presenter = value;
                    presenter.PropertyChanged += OnHighSchoolPropertyChanged;
                }
            }
        }

        void OnHighSchoolPropertyChanged(object sender, PropertyChangedEventArgs args)
        {
 	        if (args.PropertyName == "StudentBody" && presenter.StudentBody != null)
            {
                this.Groups = new ObservableCollection<StudentGroup>();
                this.Groups.Add(new StudentGroup { Title = "Male" });
                this.Groups.Add(new StudentGroup { Title = "Female" });
    
                foreach (Student student in presenter.StudentBody.Students)
                    if (student.Sex == "Male")
                        this.Groups[0].Students.Add(student);
                    else
                        this.Groups[1].Students.Add(student);
            }
        }

        public ObservableCollection<StudentGroup> Groups
        {
            set { SetProperty<ObservableCollection<StudentGroup>>(ref groups, value); }
            get { return groups; }
        }

        protected bool SetProperty<T>(ref T storage, T value,
                              [CallerMemberName] string propertyName = null)
        {
            if (object.Equals(storage, value))
                return false;

            storage = value;
            OnPropertyChanged(propertyName);
            return true;
        }

        protected void OnPropertyChanged(string propertyName)
        {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter12/GroupBySex/GroupBySex/StudentGroups.cpp
#include "pch.h"
#include "StudentGroups.h"

using namespace GroupBySex;

using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml::Data;

using namespace ElPasoHighSchool;

StudentGroups::StudentGroups()
{
    this->Groups = ref new Vector<StudentGroup^>();
}

void StudentGroups::Source::set(StudentBodyPresenter^ value)
{
    if (value != nullptr)
    {
        presenter = value;
        presenter->PropertyChanged += ref new PropertyChangedEventHandler(this, &StudentGroups::OnHighSchoolPropertyChanged);
    }
}

void StudentGroups::Groups::set(IVector<StudentGroup^>^ value)
{
    if (groups != value)
    {
        groups = value;
        OnPropertyChanged("Groups");
    }
}

void StudentGroups::OnHighSchoolPropertyChanged(Object^ sender, PropertyChangedEventArgs^ args)
{
    if (args->PropertyName == "StudentBody")
    {
        this->Groups->Clear();

        if (presenter->StudentBody != nullptr)
        {
            StudentGroup^ males = ref new StudentGroup();
            males->Title = "Male";
            this->Groups->Append(males);

            StudentGroup^ females = ref new StudentGroup();
            females->Title = "Female";
            this->Groups->Append(females);

            for each (Student^ student in presenter->StudentBody->Students)
            {
                if (student->Sex == "Male")
                    males->Students->Append(student);
                else
                    females->Students->Append(student);
            }
        }
    }
}
